internal/mic/mic.go

```diff
@@ -0,0 +1,85 @@
+package mic
+
+import (
+	"context"
+	"fmt"
+
+	kmmv1beta1 "github.com/kubernetes-sigs/kernel-module-management/api/v1beta1"
+	v1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+//go:generate mockgen -source=mic.go -package=mic -destination=mock_mic.go
+
+type ModuleImagesConfigAPI interface {
+	HandleModuleImagesConfig(ctx context.Context, name, ns string, images []kmmv1beta1.ModuleImageSpec,
```

Comment:
I think that it is better to have a function that updates one entry in the in-memory MIc object, and let the caller handle CreatOrPatch as needed.

Reply:
I think differently on this one.

I think those internal packages should be as smart as possible **as long as they stay in their scope** to reduce as max as possible responsibility from the controllers. Those controllers are getting very complicated very fast (and then we start writing helper interfaces for them).

In this case, the `MIC` interface is responsible for everything related to `ModuleImagesConfig` objects, therefore, updating the object is in its scope and it reduce a lot of functionality from the `Module` controller which should not care about how `ModuleImagesConfig`s are "internally" handled.

Reply:
The controllers usually want to update the images one at a time, since they are looping around the input and decide upon each image. Better interface would be  for controller to say: this is the image i want to query about, and let the package prepare the needed structure and find the correct entry.  Controller will look like:
loop over input:
    add image to mic if needed
commit changes into etcd

Reply:
IBut how does the controller know what image has changed (or if what was changed is metadata/imageRepoSecret)?

This is how I envision the `Module` controller: https://github.com/kubernetes-sigs/kernel-module-management/pull/993/files#diff-3e13aac4e240cd59f1a0bd375f519a14e5846cc38980d3b5018f87e133a6b56fR111-R122